contracts/src/utils/structs/checkpoints.rs

```diff
@@ -0,0 +1,463 @@
+    pub fn upper_lookup(&self, key: U96) -> U160 {
+        let len = self.length();
+        let pos = self._upper_binary_lookup(key, U256::ZERO, len);
+        if pos == U256::ZERO {
+            U160::ZERO
+        } else {
+            self._access(pos - uint!(1_U256))._value.get()
+        }
+    }
+
+    /// Returns the value in the last (most recent) checkpoint with key
+    /// lower or equal than the search key, or zero if there is none.
+    ///
+    /// This is a variant of [`Self::upper_lookup`] that is optimized to find
+    /// "recent" checkpoint (checkpoints with high keys).
+    ///
+    /// # Arguments
+    ///
+    /// * `&self` - read access to the checkpoint's state.
+    /// * `key` - Checkpoint's key to query.
+    pub fn upper_lookup_recent(&self, key: U96) -> U160 {
+        let len = self.length();
+
+        let mut low = U256::ZERO;
+        let mut high = len;
+
+        if len > uint!(5_U256) {
+            let mid = len - len.sqrt();
+            if key < self._access(mid)._key.get() {
+                high = mid;
+            } else {
+                low = mid + uint!(1_U256);
+            }
+        }
+
+        let pos = self._upper_binary_lookup(key, low, high);
+
+        if pos == U256::ZERO {
+            U160::ZERO
+        } else {
+            self._access(pos - uint!(1_U256))._value.get()
+        }
+    }
+
+    /// Returns the value in the most recent checkpoint, or zero if
+    /// there are no checkpoints.
+    ///
+    /// # Arguments
+    ///
+    /// * `&self` - read access to the checkpoint's state.
+    pub fn latest(&self) -> U160 {
+        let pos = self.length();
+        if pos == U256::ZERO {
+            U160::ZERO
+        } else {
+            self._access(pos - uint!(1_U256))._value.get()
+        }
+    }
+
+    /// Returns whether there is a checkpoint in the structure (i.g. it
+    /// is not empty), and if so, the key and value in the most recent
+    /// checkpoint.
+    /// Otherwise, [`None`] will be returned.
+    ///
+    /// # Arguments
+    ///
+    /// * `&self` - read access to the checkpoint's state.
+    pub fn latest_checkpoint(&self) -> Option<(U96, U160)> {
+        let pos = self.length();
+        if pos == U256::ZERO {
+            None
+        } else {
+            let checkpoint = self._access(pos - uint!(1_U256));
+            Some((checkpoint._key.get(), checkpoint._value.get()))
+        }
+    }
+
+    /// Returns the number of checkpoints.
+    ///
+    /// # Arguments
+    ///
+    /// * `&self` - read access to the checkpoint's state.
+    pub fn length(&self) -> U256 {
+        U256::from(self._checkpoints.len())
+    }
+
+    /// Returns checkpoint at given position.
+    ///
+    /// # Arguments
+    ///
+    /// * `&self` - read access to the checkpoint's state.
+    /// * `pos` - index of the checkpoint.
+    pub fn at(&self, pos: U32) -> Checkpoint160 {
+        unsafe { self._checkpoints.get(pos).unwrap().into_raw() }
+    }
```

Comment:
We should document that this can panic and we should document safety guarantees in a comment. If there is no way to guarantee safety, we should not use unsafe.

I think I don't understand the need for this if we have `_access`?

Reply:
Yep. I also was thinking to remove it
p.s. in solidity it was as a public api. `access` and `at` at solidity have different types of arguments (indexes) u256 and u32. And access is basically `unsafe_access`. It uses assembly and works without bound checks. I can't see at stylus functions with indexing without bound checks are public and it always be checking bounds

Reply:
I'll leave it to your judgment!

Reply:
I've left it to return tuple of (key, value) same as for `latest_checkpoint` function